cli/src/commands/chain_spec/mod.rs

```diff
@@ -0,0 +1,131 @@
+    /// Replaced the genesis raw entry with a stateRootHash to optimize
+    /// the spec size and avoid the need to calculate the genesis storage.
+    ///
+    /// This option is enabled with the `chain-spec-root-hash` feature.
+    ///
+    /// Defaults to `false`.
+    #[cfg(feature = "chain-spec")]
+    #[clap(long)]
+    state_root_hash: bool,
+    /// Remove the `codeSubstitutes` entry from the chain spec.
+    /// This is useful when wanting to store a smaller chain spec.
+    /// At this moment, the light client does not utilize this object.
+    ///
+    /// Defaults to `false`.
+    #[clap(long)]
+    remove_substitutes: bool,
+}
+
+/// Error attempting to fetch chainSpec.
+#[derive(Debug, thiserror::Error)]
+#[non_exhaustive]
+pub enum ChainSpecError {
+    /// Failed to fetch the chain spec.
+    #[error("Failed to fetch the chain spec: {0}")]
+    FetchError(#[from] fetch::FetchSpecError),
+
+    #[cfg(feature = "chain-spec")]
+    /// The provided chain spec is invalid.
+    #[error("Error while parsing the chain spec: {0})")]
+    ParseError(String),
+
+    #[cfg(feature = "chain-spec")]
+    /// Cannot compute the state root hash.
+    #[error("Error computing state root hash: {0})")]
+    ComputeError(String),
+
+    /// Other error.
+    #[error("Other: {0})")]
+    Other(String),
+}
+
+#[cfg(feature = "chain-spec")]
+fn compute_state_root_hash(spec: &Value) -> Result<[u8; 32], ChainSpecError> {
+    let chain_spec = smoldot::chain_spec::ChainSpec::from_json_bytes(spec.to_string().as_bytes())
+        .map_err(|err| ChainSpecError::ParseError(err.to_string()))?;
+
+    let genesis_chain_information = chain_spec.to_chain_information().map(|(ci, _)| ci);
+
+    let state_root = match genesis_chain_information {
+        Ok(genesis_chain_information) => {
+            let header = genesis_chain_information.as_ref().finalized_block_header;
+            *header.state_root
+        }
+        // From the smoldot code this error is encountered when the genesis already contains the
+        // state root hash entry instead of the raw entry.
+        Err(smoldot::chain_spec::FromGenesisStorageError::UnknownStorageItems) => *chain_spec
+            .genesis_storage()
+            .into_trie_root_hash()
+            .ok_or_else(|| {
+                ChainSpecError::ParseError(
+                    "The chain spec does not contain the proper shape for the genesis.raw entry"
+                        .to_string(),
+                )
+            })?,
+        Err(err) => return Err(ChainSpecError::ComputeError(err.to_string())),
+    };
+
+    Ok(state_root)
+}
+
+pub async fn run(opts: Opts, output: &mut impl Write) -> color_eyre::Result<()> {
+    let url = opts.url;
+
+    let mut spec = fetch::fetch_chain_spec(url).await?;
+
+    let mut output: Box<dyn Write> = match opts.output_file {
+        Some(path) => Box::new(std::fs::File::create(path)?),
+        None => Box::new(output),
+    };
+
+    #[cfg(feature = "chain-spec")]
```

Comment:
I am a bit torn about the `chain-spec` feature. It is the first feature flag introduced to the subxt CLI tool and having feature flags on the CLI tool means that people can have different experiences using the subxt CLI tool, depending on which features they installed it with. However I understand that smoldot is quite a big dependency that we do not want to expose unnecessarily for all users.
We could also guard the entire `chainspec` subcommand behind the `"chain-spec"` feature flag, then we do not need so many `#[cfg(feature = "chain-spec")]` instances throughout the code. But if we keep it like this, I would probably rename it to something more clear like `chain-spec-pruning`.

Reply:
Would be nice to get rid of the feature completely  :)